R/GeospatialFunctions.R

```diff
@@ -0,0 +1,612 @@
+    
+    stop("You data has already been joined with ATTAINS data.")
+    
+  } else if(nrow(data) == 0){
+    
+    print("Your Water Quality Portal dataframe has no observations. Returning an empty dataframe with empty ATTAINS features.")
+    
+    # if no WQP observations, return a modified `data` with empty ATTAINS-related columns:
+    no_WQP_data <- data %>%
+      dplyr::mutate("index", "ATTAINS.organizationid" = NA, "ATTAINS.submissionid" = NA, "ATTAINS.hasprotectionplan" = NA,
+                    "ATTAINS.assessmentunitname" = NA, "ATTAINS.nhdplusid" = NA, "ATTAINS.tas303d" = NA,                                                  
+                    "ATTAINS.isthreatened" = NA, "ATTAINS.state" = NA, "ATTAINS.on303dlist" = NA,                                               
+                    "ATTAINS.organizationname" = NA, "ATTAINS.region" = NA, "ATTAINS.Shape_Length" = NA,                                             
+                    "ATTAINS.reportingcycle" = NA, "ATTAINS.assmnt_joinkey" = NA, "ATTAINS.hastmdl" = NA,                                                  
+                    "ATTAINS.orgtype" = NA, "ATTAINS.permid_joinkey" = NA, "ATTAINS.catchmentistribal" = NA,                                        
+                    "ATTAINS.ircategory" = NA, "ATTAINS.waterbodyreportlink" = NA, "ATTAINS.assessmentunitidentifier" = NA,                                 
+                    "ATTAINS.overallstatus" = NA, "ATTAINS.isassessed" = NA, "ATTAINS.isimpaired" = NA,                                               
+                    "ATTAINS.has4bplan" = NA, "ATTAINS.huc12" = NA, "ATTAINS.hasalternativeplan" = NA,                                      
+                    "ATTAINS.visionpriority303d" = NA, "ATTAINS.areasqkm" = NA, "ATTAINS.catchmentareasqkm" = NA,                                       
+                    "ATTAINS.catchmentstatecode" = NA, "ATTAINS.catchmentresolution" = NA, "ATTAINS.Shape_Area" = NA)
+    
+    if(return == TRUE){
+      ATTAINS_catchments <- NULL
+      ATTAINS_lines <- NULL
+      ATTAINS_points <- NULL
+      ATTAINS_polygons <- NULL
+      
+      return(list("TADA_with_ATTAINS" = no_WQP_data,
+                  "ATTAINS_catchments" = ATTAINS_catchments,
+                  "ATTAINS_points" = ATTAINS_points,
+                  "ATTAINS_lines" = ATTAINS_lines,
+                  "ATTAINS_polygons" = ATTAINS_polygons))} else {return(no_WQP_data)}
+    
+  }
+
+suppressMessages(suppressWarnings({
+  
+  sf::sf_use_s2(FALSE)
+  
+  # If data is already spatial, just make sure it is in the right CRS
+  # and add unique WQP ID for identifying obs with more than one ATTAINS assessment unit
+  
+  if (!is.null(data) & inherits(data, "sf")) {
+    if(sf::st_crs(data)$epsg != 4326){
+      TADA_DataRetrieval_data <- data %>%
+        sf::st_transform(4326) %>%
+        tibble::rowid_to_column(var = "index")
+    } else {
```

Comment:
This line makes me wonder about the `!is.null(data)` check on L292: Is that check redundant at all with L253's ` } else if(nrow(data) == 0){`? If it's not, do you think it would make sense to make L297 `} else if(!is.null(data) & !inherits(data, "sf")`?